Lib/functools.py

```diff
@@ -753,76 +753,78 @@ def singledispatch(func):
     implementations can be registered using the register() attribute of the
     generic function.
     """
-    # There are many programs that use functools without singledispatch, so we
-    # trade-off making singledispatch marginally slower for the benefit of
-    # making start-up of such applications slightly faster.
-    import types, weakref
-
-    registry = {}
-    dispatch_cache = weakref.WeakKeyDictionary()
-    cache_token = None
-
-    def dispatch(cls):
-        """generic_func.dispatch(cls) -> <function implementation>
-
-        Runs the dispatch algorithm to return the best available implementation
-        for the given *cls* registered on *generic_func*.
-
-        """
-        nonlocal cache_token
-        if cache_token is not None:
-            current_token = get_cache_token()
-            if cache_token != current_token:
-                dispatch_cache.clear()
-                cache_token = current_token
-        try:
-            impl = dispatch_cache[cls]
-        except KeyError:
+    if func is None:
+        return partial(singledispatch, arg=arg)
+    else:
+        # There are many programs that use functools without singledispatch, so we
+        # trade-off making singledispatch marginally slower for the benefit of
+        # making start-up of such applications slightly faster.
+        import types, weakref
+        registry = {}
+        dispatch_cache = weakref.WeakKeyDictionary()
+        cache_token = None
+
+        def dispatch(cls):
+            """generic_func.dispatch(cls) -> <function implementation>
+
+            Runs the dispatch algorithm to return the best available implementation
+            for the given *cls* registered on *generic_func*.
+
+            """
+            nonlocal cache_token
+            if cache_token is not None:
+                current_token = get_cache_token()
+                if cache_token != current_token:
+                    dispatch_cache.clear()
+                    cache_token = current_token
             try:
-                impl = registry[cls]
+                impl = dispatch_cache[cls]
```

Comment:
This is the diff being weird, I did not change this. The block and main part of singledispatch was indented.